### PR TITLE
[IMP] Prevent clicking navbar buttons while starting the editor

### PR DESCRIPTION
**Pull request by tdecaluwe**
When visiting a page with enable_editor set to true, the web editor will be automatically started. However, while the editor is loading, the end user can still click other buttons in the navbar.

For example, some changes in the web editor require a save and reload of the current page being edited. After saving, the same page will be reloaded with enable_editor.

This PR blocks any clicks while the editor is being loaded.

task-2607755

### CODE REVIEW

File: addons/website/static/src/js/menu/edit.js [hunk 1]
```diff
@@ -72,9 +72,18 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
     start: function () {
         var def = this._super.apply(this, arguments);
 
+        let navbar = $('nav#oe_main_menu_navbar').get(0);
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/edit.js:75]: - const
- move it inside the if, don't define it outside if you don't need it
- suffix by `El`

File: addons/website/static/src/js/menu/edit.js [hunk 2]
```diff
@@ -72,9 +72,18 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+        let stopPropagation = function (event) {
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/edit.js:76]: - const
- move it inside the if, don't define it outside if you don't need it
- probably es6 arrow

File: addons/website/static/src/js/menu/edit.js [hunk 3]
```diff
@@ -72,9 +72,18 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+            event.stopPropagation();
+        };
+
         // If we auto start the editor, do not show a welcome message
         if (this._editorAutoStart) {
-            return Promise.all([def, this._startEditMode()]);
+            navbar.addEventListener('click', stopPropagation, {capture: true});
+            return Promise.all([def, this._startEditMode()]).then(function (result) {
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/edit.js:83]: can't you `await` + put start in async instead of then?
If you can't, at least use es6 arrow

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Wouldn't it actually be better UX to show a spinner on the whole screen to prevent anything to be done?
This will indeed work, the user will not be able to click anywhere but he won't really especially figure why.. (because there is something loading)
If we show a spinner, he will know something is being processed

File: addons/website/static/src/js/menu/edit.js (hunk 1 above)
Comment (rdeodoo) [reply addons/website/static/src/js/menu/edit.js:75]: And don't use jquery to get the element if you are going to straigh away convert it to DOM

File: addons/website/static/src/js/menu/edit.js [hunk 4]
```diff
@@ -72,9 +72,18 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
 
+        let navbar = $('nav#oe_main_menu_navbar').get(0);
+        let stopPropagation = function (event) {
+            event.stopPropagation();
+        };
+
         // If we auto start the editor, do not show a welcome message
         if (this._editorAutoStart) {
-            return Promise.all([def, this._startEditMode()]);
+            navbar.addEventListener('click', stopPropagation, {capture: true});
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/edit.js:82]:
```suggestion
            navbar.addEventListener('click', stopPropagation, true);
```

**Review by rdeodoo — COMMENTED** [review]
(I think it should be a spinner but only considering the current PR state:)
I would have personnally used jquery on click method with a namespaced click event, then would have unbound it with just `.off('click.xxx')` -> no need to store the click on a variable
it would make even more sense since you used jquery to get the element (which is wrong here)
But I think we prefer to get rid of jquery so.. I don't want to spread incorrect advice on this one, @qsm-odoo ?

**Edit**: in this case, since jquery doesn't allow the use of `capture`, **it won't work** anyway

**Comment by tdecaluwe** [comment]
@rdeodoo The spinner is definitely an option here, however I'm waiting for @pierre-pde to clear up the specs. I also noticed that, since the editor widget is loaded after the navbar, there is a very small window that the buttons are still clickable... Up until `start` is called on the editor widget.

The `capture` is indeed the reason for not using jquery event handlers in this case.

**Comment by rdeodoo** [comment]
Ok for capture/jquery, about the spinner I am not sure it require a spec cleanup from @pierre-pde to chose between a visible spinner telling the user something is going on, or a silent click prevent

> there is a very small window

Are we talking about some ms? That would not be something to fix/improve, generally it can take up to a few second before loading the editor, so the user as some whole seconds to click on elements since he doesnt know something is happening,  which your PR is fixing (with the spiner or with your current state).
But if we are talking about fixing the missing few ms before your fix, that's not something to consider IMO..

File: addons/website/static/src/js/menu/edit.js (hunk 4 above)
Comment (tdecaluwe) [reply addons/website/static/src/js/menu/edit.js:82]: Ok, I like the options form better though.

Comment (rdeodoo) [reply addons/website/static/src/js/menu/edit.js:82]: You can keep it like that it is a minor remark :+1: , don't really mind it..
I just checked on our codebase how we were writing it: with the direct form. Plus saw the doc that suggest a direct form when only setting capture to true

File: addons/website/static/src/js/menu/edit.js [hunk 5]
```diff
@@ -63,18 +63,28 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
         url.searchParams.delete('enable_editor');
         url.searchParams.delete('with_loader');
         window.history.replaceState({}, null, url);
+
+        if (this._editorAutoStart) {
+            $('body').block({overlayCSS: {
+                backgroundColor: '#000',
+                opacity: 0.5,
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/edit.js:70]: The 0.5 opacity seems weird visually, it is too much.
I would just keep 0 as in the commit I mentionned you  https://github.com/odoo/odoo/commit/eb3ed89f8440050fc32f9a7b29fb2b6ab8fd6b59 so visually only the cursor is spinning (light effect vs spin + applying an opacity on top of the whole screen)).

File: addons/website/static/src/js/menu/edit.js [hunk 6]
```diff
@@ -63,18 +63,28 @@ var EditPageMenu = websiteNavbarData.WebsiteNavbarActionWidget.extend({
...
+            }, message: false});
+        }
     },
     /**
      * Auto-starts the editor if necessary or add the welcome message otherwise.
      *
      * @override
      */
-    start: function () {
+    start: async function () {
```
Comment (rdeodoo) [inline addons/website/static/src/js/menu/edit.js:80]: Minor: Since you edit this line
```suggestion
    async start() {
```

**Review by rdeodoo — CHANGES_REQUESTED** [review]
LGTM, just a small remark about the opacity (and a minor one that you can ignore if you want about the start)

Also, why not 15.0?

File: addons/website/static/src/js/menu/edit.js (hunk 5 above)
Comment (tdecaluwe) [reply addons/website/static/src/js/menu/edit.js:70]: I added the opacity to make it visually more clear that the editor is loading. Either way way is fine by me though, I don't feel I'm the right person to argue about this :) Consistency is probably just as important?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@tdecaluwe

**Review by qsm-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**